src/libosd/cl_dem_uart.c

```diff
@@ -0,0 +1,110 @@
+/* Copyright 2018 The Open SoC Debug Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <osd/cl_dem_uart.h>
+
+#include <assert.h>
+#include <osd/osd.h>
+#include <stdio.h>
+#include <unistd.h>
+#include "osd-private.h"
+
+static bool is_dem_uart_module(struct osd_hostmod_ctx *hostmod_ctx,
+                          unsigned int dem_uart_di_addr)
+{
+    osd_result rv;
+
+    struct osd_module_desc desc;
+
+    rv = osd_hostmod_mod_describe(hostmod_ctx, dem_uart_di_addr, &desc);
+    if (OSD_FAILED(rv)) {
+        return false;
+    }
+
+    if (desc.vendor != OSD_MODULE_VENDOR_OSD ||
+        desc.type != OSD_MODULE_TYPE_STD_DEM_UART || desc.version != 0) {
+        return false;
+    }
+
+    return true;
+}
+
+API_EXPORT
+osd_result osd_cl_dem_uart_get_desc(struct osd_hostmod_ctx *hostmod_ctx,
+									unsigned int dem_uart_di_addr,
+									struct osd_dem_uart_desc *dem_uart_desc)
+{
+    assert(hostmod_ctx);
+    assert(dem_uart_desc);
+
+    if (!is_dem_uart_module(hostmod_ctx, dem_uart_di_addr)) {
+    	return OSD_ERROR_WRONG_MODULE;
+    }
+
+    dem_uart_desc->diaddr = dem_uart_di_addr;
+
+    return OSD_OK;
+}
+
+API_EXPORT
+osd_result osd_cl_dem_uart_receive_event(void *arg, struct osd_packet *pkg)
+{
+    assert(arg &&
+           "You need to give an event_handler_arg of type "
+           "struct osd_dem_uart_event_handler in osd_hostmod_new()");
+    assert(pkg);
+
+    struct osd_dem_uart_event_handler *handler = arg;
+
+    struct osd_dem_uart_event *ev = calloc(1, sizeof(*ev));
+    if (!ev) {
+    	printf("malloc() failed\n");
+    }
```

Comment:
We assert in out of memory conditions, please see https://opensocdebug.readthedocs.io/projects/osd-sw/en/latest/02_developer/api/libosd/errorhandling.html#handling-memory-allocation-errors for a longer discussion.